support-models/src/main/scala/com/gu/support/encoding/DiscriminatedType.scala

```diff
@@ -0,0 +1,121 @@
+package com.gu.support.encoding
+
+import cats.implicits._
+import io.circe.Decoder.Result
+import io.circe.generic.decoding.DerivedDecoder
+import io.circe.generic.encoding.DerivedAsObjectEncoder
+import io.circe.generic.semiauto.{deriveDecoder, deriveEncoder}
+import io.circe.syntax._
+import io.circe._
+import shapeless.Lazy
+
+import scala.reflect.ClassTag
+
+object DiscriminatedType {
+
+  val defaultDiscriminatorFieldName = "_type"
+
+  def getSingle[A](list: List[A], message: String): A = list match {
+    case Nil => throw new RuntimeException(s"no codec for: $message")
+    case a :: Nil => a
+    case many => throw new RuntimeException(s"duplicate codec for: $message: $many")
+  }
+
+  def encoder[TOPLEVEL](allCodecs: List[VariantEncoder[_ <: TOPLEVEL]]): Encoder[TOPLEVEL] = {
+    if (allCodecs.groupBy(_.discriminatorFieldName).size != 1)
+      throw new RuntimeException("coding error - non matching discriminator names")
```

Comment:
after the refactor, a few things that would have been compile errors are now runtime exceptions

Reply:
Could we not have just changed the constructor of the `DiscriminatedType` class to have a default? Eg.
```
class DiscriminatedType[TOPLEVEL](discriminatorFieldName: String = "__type")
```

Reply:
we could, but I don't think we actually want or need a default, that would only be a tradeoff to get better syntax.

my original point was purely about the syntax looking a bit weird, but we could only improve it by standardising the discriminator field name.